Assets/MixedRealityToolkit.Services/InputSystem/Utilities/ScaleMeshEffect.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Microsoft.MixedReality.Toolkit.Input.Utilities
+{
+    /// <summary>
+    /// On Unity UI components the unity_ObjectToWorld matrix is not the transformation matrix of the local 
+    /// transform the Graphic component lives on, but that of it's parent Canvas. Many MRTK/Standard shader 
+    /// effects require object scale to be known. To solve this issue the ScaleMeshEffect will store scaling 
+    /// information into UV channel attributes during UI mesh construction. Ideally we would store the scale 
+    /// in one attribute but UGUI only supports two scalers per attribute (even in the tangent attribute).
+    /// </summary>
+    [RequireComponent(typeof(RectTransform))]
+    public class ScaleMeshEffect : BaseMeshEffect
+    {
+        /// <summary>
+        /// Enforces the parent canvas uses UV channel attributes.
+        /// </summary>
+        protected override void Awake()
+        {
+            base.Awake();
+
+            // Make sure the parent canvas is configured to use UV channel attributes for scaling data.
+            var canvas = GetComponentInParent<Canvas>();
+
+            if (canvas != null)
+            {
+                canvas.additionalShaderChannels |= AdditionalCanvasShaderChannels.TexCoord2;
+                canvas.additionalShaderChannels |= AdditionalCanvasShaderChannels.TexCoord3;
+            }
+        }
+
+        /// <summary>
+        /// Stores scaling information into UV channel attributes during UI mesh construction.
+        /// </summary>
+        /// <param name="vh">The vertex helper to populate with new vertex data.</param>
+        public override void ModifyMesh(VertexHelper vh)
+        {
+            var rectTransform = transform as RectTransform;
+            var scale = new Vector2(rectTransform.rect.width * rectTransform.localScale.x,
+                                    rectTransform.rect.height * rectTransform.localScale.y);
+
+            var canvas = GetComponentInParent<Canvas>();
+            var depth = new Vector2((canvas ? (1.0f / canvas.transform.lossyScale.z) : 1.0f) * rectTransform.localScale.z, 
+                                    -1.0f);
```

Comment:
I see.  Here's where you're setting it to the -1 you mentioned in the shader.  Maybe add a comment here saying that that's why you're setting it to -1 here.

Reply:
Good idea! I’ll add a comment expanding what is going on here and in the shader. The way the data is packed is a little obuscated.